plugins/generic_plugin.c

```diff
@@ -0,0 +1,206 @@
+}
+
+static int compare_gain(const char *a, const char *b) __attribute__((unused));
+static int compare_gain(const char *a, const char *b)
+{
+	double val_a, val_b;
+	sscanf(a, "%lf", &val_a);
+	sscanf(b, "%lf", &val_b);
+
+	if (val_a < val_b)
+		return -1;
+	else if(val_a > val_b)
+		return 1;
+	else
+		return 0;
+}
+
+static void save_widget_value(GtkWidget *widget, struct iio_widget *iio_w)
+{
+	iio_w->save(iio_w);
+}
+
+static void make_widget_update_signal_based(struct iio_widget *widgets,
+		unsigned int num_widgets)
+{
+	char signal_name[25];
+	unsigned int i;
+
+	for (i = 0; i < num_widgets; i++) {
+		if (GTK_IS_CHECK_BUTTON(widgets[i].widget))
+			sprintf(signal_name, "%s", "toggled");
+		else if (GTK_IS_TOGGLE_BUTTON(widgets[i].widget))
+			sprintf(signal_name, "%s", "toggled");
+		else if (GTK_IS_SPIN_BUTTON(widgets[i].widget))
+			sprintf(signal_name, "%s", "value-changed");
+		else if (GTK_IS_COMBO_BOX_TEXT(widgets[i].widget))
+			sprintf(signal_name, "%s", "changed");
+		else
+			printf("unhandled widget type, attribute: %s\n", widgets[i].attr_name);
+
+		if (GTK_IS_SPIN_BUTTON(widgets[i].widget) &&
+		    widgets[i].priv_progress != NULL) {
+			iio_spin_button_progress_activate(&widgets[i]);
+		} else {
+			g_signal_connect(G_OBJECT(widgets[i].widget), signal_name,
+					 G_CALLBACK(save_widget_value), &widgets[i]);
+		}
+	}
+}
+
+static bool generic_identify(const struct osc_plugin *plugin)
+{
+	return 1;
+}
+
+static GSList * get_dac_dev_names(void)
+{
+	GSList *node;
+	GSList *dac_dev_names = NULL;
+	struct osc_plugin *plugin;
+
+	for (node = plugin_list; node; node = g_slist_next(node)) {
+		plugin = node->data;
+		if (plugin->get_dac_dev_names)
+			dac_dev_names = g_slist_concat (dac_dev_names, plugin->get_dac_dev_names());
+	}
+
+	return dac_dev_names;
+}
+
+static gint compare_func(gconstpointer a, gconstpointer b)
+{
+	const char *pa = a, *pb = b;
+
+	return strcmp(pa, pb);
+}
+
+static GtkWidget * generic_init(struct osc_plugin *plugin, GtkWidget *notebook,
+				const char *ini_fn)
+{
+	GtkBuilder *builder;
+	GtkWidget *generic_panel;
+	GtkWidget *dds_container;
+	GtkWidget *dds_vbox;
+	bool generic_en = false;
+	int i, dev_count;
+	GSList * dac_dev_names;
+	const char *crt_dev_name;
+
+	ctx = osc_create_context();
+	if (!ctx)
+		return NULL;
+
+	builder = gtk_builder_new();
+	if (osc_load_glade_file(builder, "generic_plugin") < 0)
+		return NULL;
+
+	generic_panel = GTK_WIDGET(gtk_builder_get_object(builder, "generic_panel"));
+	dds_vbox = GTK_WIDGET(gtk_builder_get_object(builder, "dds_vbox"));
+	dac_dev_names = get_dac_dev_names();
+	dev_count = iio_context_get_devices_count(ctx);
+
+	for (i = 0; i < dev_count; i++) {
+		dac = iio_context_get_device(ctx, i);
+		dac_tx_manager = dac_data_manager_new(dac, NULL, ctx);
+		if (!dac_tx_manager)
+			continue;
+
+		crt_dev_name = iio_device_get_name(dac);
+		if (g_slist_find_custom(dac_dev_names, crt_dev_name, compare_func))
+			continue;
```

Comment:
There's a memory leak happening here: at this point there is memory allocated for dac_tx_manager and not freed anywhere.

Reply:
Fix memory leak.